### PR TITLE
[SD-488] [SD-779][SD-717] Segment/GA with turbolinks

### PR DESCRIPTION
**Pull request by Punkbooster**

### CODE REVIEW

File: app/views/spree/shared/trackers/segment/_cart_viewed.js.erb [hunk 1]
```diff
@@ -1,22 +1,23 @@
...
-      <% end %>
-      ]
-    });
+  window.addEventListener('turbolinks:load', function() {
+    if (typeof analytics !== 'undefined') {
+      analytics.track('Cart Viewed', {
+        cart_id: '<%= @order.number %>',
+        products: [
+        <% @order.line_items.each_with_index do |line_item, index| %>
+          <%= product_for_segment(line_item, position: index+1, quantity: line_item.quantity, image: line_item.variant.images.first) %>,
```
Comment (damianlegawiec) [inline app/views/spree/shared/trackers/segment/_cart_viewed.js.erb]:
```suggestion
          <%= product_for_segment(line_item, position: index+1, quantity: line_item.quantity, image: default_image_for_product_or_variant(line_item.variant)) %>,
```

File: app/views/spree/shared/trackers/segment/_cart_viewed.js.erb [hunk 2]
```diff
@@ -1,22 +1,23 @@
...
-      <% end %>
-      ]
-    });
-
-  }
+      analytics.page('Cart Viewed', {
+        cart_id: '<%= @order.number %>',
+        products: [
+        <% @order.line_items.each_with_index do |line_item, index| %>
+          <%= product_for_segment(line_item, position: index+1, quantity: line_item.quantity, image: line_item.variant.images.first) %>,
```
Comment (damianlegawiec) [inline app/views/spree/shared/trackers/segment/_cart_viewed.js.erb]:
```suggestion
          <%= product_for_segment(line_item, position: index+1, quantity: line_item.quantity, image: default_image_for_product_or_variant(line_item.variant) %>,
```

File: app/views/spree/shared/trackers/segment/_checkout_step_viewed.js.erb [hunk 3]
```diff
@@ -1,12 +1,14 @@
...
-    analytics.track('Checkout Step Viewed', {
-      checkout_id: '<%= @order.number %>',
-      step: '<%= (@order.checkout_steps.index(@order.state) + 1) %>'
-    });
-    analytics.page('Checkout Step Viewed', {
-      checkout_id: '<%= @order.number %>',
-      step: '<%= (@order.checkout_steps.index(@order.state) + 1) %>'
-    });
-  }
+  window.addEventListener('turbolinks:load', function() {
```
Comment (damianlegawiec) [inline app/views/spree/shared/trackers/segment/_checkout_step_viewed.js.erb]:
```suggestion
  <% if @order.present? && !@order.completed? %>
  window.addEventListener('turbolinks:load', function() {
```

File: app/views/spree/shared/trackers/segment/_order_complete.js.erb [hunk 4]
```diff
@@ -1,48 +1,49 @@
...
+          total: <%= @order.total %>,
+          shipping: <%= @order.shipment_total %>,
+          tax: <%= @order.additional_tax_total %>,
+          discount: <%= @order.promo_total %>,
+          coupon: '<%= @order.promo_code %>',
+          currency: '<%= @order.currency %>',
+          products: [
+          <% @order.line_items.each do |line_item| %>
+            {
+              product_id: '<%= line_item.product.id %>',
```
Comment (damianlegawiec) [inline app/views/spree/shared/trackers/segment/_order_complete.js.erb]:
```suggestion
              product_id: '<%= line_item.product_id %>',
```

File: app/views/spree/shared/trackers/segment/_order_complete.js.erb [hunk 5]
```diff
@@ -1,48 +1,49 @@
...
+              sku: '<%= line_item.variant.sku %>',
```
Comment (damianlegawiec) [inline app/views/spree/shared/trackers/segment/_order_complete.js.erb]:
```suggestion
              sku: '<%= line_item.sku %>',
```

File: app/views/spree/shared/trackers/segment/_order_complete.js.erb [hunk 6]
```diff
@@ -1,48 +1,49 @@
...
+              name: '<%= line_item.product.name %>',
```
Comment (damianlegawiec) [inline app/views/spree/shared/trackers/segment/_order_complete.js.erb]:
```suggestion
              name: '<%= line_item.name %>',
```

File: app/views/spree/shared/trackers/segment/_product_list_filtered.js.erb [hunk 7]
```diff
@@ -1,32 +1,33 @@
...
+      if (typeof analytics !== 'undefined') {
+        analytics.track('Product List Filtered', {
+          filters: [
+            <% params[:search].to_unsafe_h.each do |type, value| %>
+              <%= { type: type, value: value } .to_json.html_safe %>,
+            <% end %>
+          ],
+          products: [
+          <% @products.each_with_index do |product, index| %>
+            <%= product_for_segment(product, position: index+1, image: product.images.first) %>,
```
Comment (damianlegawiec) [inline app/views/spree/shared/trackers/segment/_product_list_filtered.js.erb]:
```suggestion
            <%= product_for_segment(product, position: index+1, image: default_image_for_product_or_variant(product)) %>,
```

File: app/views/spree/shared/trackers/segment/_product_list_viewed.js.erb [hunk 8]
```diff
@@ -1,22 +1,23 @@
...
-      <% end %>
-      ]
-    });
+  window.addEventListener('turbolinks:load', function() {
+    if (typeof analytics !== 'undefined') {
+      analytics.track('Product List Viewed', {
+        category: '<%= @taxon.try(:name) %>',
+        products: [
+        <% products.each_with_index do |product, index| %>
+          <%= product_for_segment(product, position: index+1, image: product.images.first) %>,
```
Comment (damianlegawiec) [inline app/views/spree/shared/trackers/segment/_product_list_viewed.js.erb]:
```suggestion
          <%= product_for_segment(product, position: index+1, image: default_image_for_product_or_variant(product)) %>,
```

File: app/views/spree/shared/trackers/segment/_product_list_viewed.js.erb [hunk 9]
```diff
@@ -1,22 +1,23 @@
...
-      <% end %>
-      ]
-    });
-
-  }
+      analytics.page('Product List Viewed', {
+        category: '<%= @taxon.try(:name) %>',
+        products: [
+        <% products.each_with_index do |product, index| %>
+          <%= product_for_segment(product, position: index+1, image: product.images.first) %>,
```
Comment (damianlegawiec) [inline app/views/spree/shared/trackers/segment/_product_list_viewed.js.erb]:
```suggestion
          <%= product_for_segment(product, position: index+1, image: default_image_for_product_or_variant(product)) %>,
```

**Comment by PSzyszkaSpark** [comment]
After selecting `place order` button on `checkout/confirm` step I get this error: `Uncaught SyntaxError: Unexpected token '{'` in console . Please verify it before merging.
Screen Shot 2020-02-24 at 12 28 22

**Comment by runlevel5** [comment]
@Punkbooster could you please update the description on what this PR is about?

**Pull request closed without merging**